src/uint256.cpp

```diff
@@ -18,39 +18,31 @@ std::string base_blob<BITS>::GetHex() const
 }
 
 template <unsigned int BITS>
-void base_blob<BITS>::SetHex(const char* psz)
+void base_blob<BITS>::SetHex(std::string_view str)
 {
     std::fill(m_data.begin(), m_data.end(), 0);
 
-    // skip leading spaces
-    while (IsSpace(*psz))
-        psz++;
-
-    // skip 0x
-    if (psz[0] == '0' && ToLower(psz[1]) == 'x')
-        psz += 2;
+    str = util::TrimStringView(str);
+    str = util::RemovePrefixView(str, "0x");
 
     // hex string to uint
     size_t digits = 0;
-    while (::HexDigit(psz[digits]) != -1)
-        digits++;
+    for (const char c : str) {
+        if (::HexDigit(c) == -1)
+            break;
+        ++digits;
+    }
```

Comment:
It is random access, but it's also contiguous, allowing for the safe range-`for`. If I had to do it again I might have lifted the `break;` up to the same line as the `if`. If I need to retouch and others support that change I might do that to make it slightly more terse like yours.

Reply:
You can of course redo it, I didn't review your change for us to ponder on what could have been...

Reply:
> You can of course redo it, I didn't review your change for us to ponder on what could have been...

I think the while loop is a slightly better but range-for is also ok.

I still do think as long as you are going to update the loop to respect the length of the input (instead of blowing past it) it would be better update the loop to respect the length of the output as well, and not pointlessly read more data than can actually be used, as suggested https://github.com/bitcoin/bitcoin/pull/30436#discussion_r1683086730.

Not very important though, especially if this function will be deleted later.

Reply:
> I still do think as long as you are going to update the loop to respect the length of the input (instead of blowing past it) it would be better update the loop to respect the length of the output as well, and not pointlessly read more data than can actually be used, as suggested https://github.com/bitcoin/bitcoin/pull/30436#discussion_r1683086730.

I tried making the loop counting `digits` actually respect the output length (`WIDTH*2` here) as well.
```C++
    for (const char c : trimmed) {
        if (digits == WIDTH*2 || ::HexDigit(c) == -1) break;
        ++digits;
    }
```
This uncovered some weird behavior. Suddenly this check started failing:
```C++
    BOOST_CHECK_EQUAL(R1S, uint160S(R1ArrayHex));
```
Because of how the following loop in `SetHex()` parses digits from the end, it turns out that without the limit on `digits`, **if we were given more hex digits than fit, we would read the rightmost ones rather than the leftmost ones.** :exploding_head:

(The behavior of reading the rightmost hex digits predates this PR, and I'd rather not go down that rabbit hole, as the function will hopefully be removed - https://github.com/bitcoin/bitcoin/pull/30436#discussion_r1683226283).